rfcs/007-background-workers.md

```diff
@@ -0,0 +1,299 @@
+happens via the database. Workers can start and stop at any time. Workers
+monitor each other for liveliness and in case some workers abruptly terminate,
+all the jobs of a dead worker are re-enqueued into the global pending queue.
+
+## Requirements Language
+
+[NOTE]: # ( Do not alter the section below. Follow its instructions. )
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+---
+
+`Job`: An unit of work, identified by a `JobId` and also having a `JobType`.
+
+`Job table`: A subspace holding the list of jobs indexed by `JobId`.
+
+`Pending queue`: A queue of jobs which are ready to run. Workers may pick jobs
+from this queue and start running them.
+
+`Active jobs`: A subspace holding the list of jobs currently run by a
+particular worker.
+
+`Worker`: A job execution unit. Workers could be individual processes or groups
+of processes running on remote nodes.
+
+`Health key` : A key that the worker periodically updates with a timestamp to
+indicate that they are "alive" and ready to process jobs.
+
+`Versionstamp`: a 12 byte, unique, monotonically (but not sequentially)
+increasing value for each committed transaction.
+
+---
+
+# Detailed Description
+
+## Data Model
+
+The main job table:
+ - `("couch_workers", "jobs", JobType, JobId) = (JobState, WorkerId, Priority, CancelReq, HealthTimeout, JobInfo, JobOps)`
+
+Pending queue:
+ - `("couch_workers", "pending", JobType, Priority, JobId) = ""`
+
+Active jobs:
+ - `("couch_workers", "active", WorkerType, Worker, JobId) = (JobState, JobInfo, JobOpts)`
+
+Worker registration and health:
+ - `("couch_workers", WorkerType, "workers_vs") = Versionstamp`
+ - `("couch_workers", WorkerType, "workers", WorkerId) = WorkerOpts`
+ - `("couch_workers", WorkerType, "health", WorkerId) = HealthCounter`
+
+### Data model fields
+
+- `JobType`: The job type. It MAY be ``replication`` or ``indexing``, however,
+  specific types are not in scope of this document.
+
+- `JobId` : MUST uniquely identify a job in a cluster. It MAY be a `UUID` or
+  derived from some other job parameters. In case the same job is submitted
+  multiple times, it is up to the job creation and type-specific worker logic
+  to cancel, re-start or update jobs.
+
+- `JobState` : MUST be one of `pending`, `running`, or `completed`
+
+- `Priority` : MAY be any key which allows sorting jobs, such as a timestamp,
+  or a tag.
+
+- `WorkerId` : MUST uniquely identify a worker in a cluster. It MAY be
+  generated each time the worker is restarted, or it could be persisted across
+  worker restarts, as long as it is unique across the whole cluster.
+
+- `CancelReq` : A boolean flag indicating a request to the worker to cancel the
+  job.
+
+- `JobOps` : Type-specific job options. Represented as a JSON object. This MAY
```

Comment:
how is the JSON marshalled to fdb?

Reply:
As a binary. In Erlang it would be jiffy:encode(...)